CHERI-UPDATING.md

```diff
@@ -79,6 +79,155 @@ Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
+only eligible commits and to cause the commit date to be displayed we
+use:
+```
+$ git log --first-parent --format=fuller upstream/main
+```
+We can then use the pager's search function to search for the
+appropriate date (e.g.. `Commit.*Fri Feb 12`). It's a good idea to check
+a few commits up or down to make sure the commit isn't at an obviously
+bad point. Then we create a tag for the commit:
+```
+$ git tag -m "FreeBSD main DD Month YYYY" freebsd-main-<YYYYMMDD> <hash>
+```
+
+#### Performing the merge
+First, create a merge branch from an update to date `dev` checkout:
+```
+$ git checkout -b merge-freebsd-<YYYYMMDD> dev
+```
+Then start the merge:
+```
+$ mergify start freebsd-main-<YYYYMMDD>
+```
+The mergify command will start merging each change one at a time until
+it hits a conflict or a pause point (e.g., a path in the merged commit
+matches the regex in [.mergify_pause_paths].) After inspecting the
+change or resolving the conflict, continue with:
+```
+mergify continue
+```
+
+For ease of conflict resolution, `mergify resolve` opens each file
+containing conflict files in `EDITOR` in turn. `mergify autoadd` adds
+all conflicted files that no longer contain conflict markers to the
+index. It does not add files that were not in the original merge, they
+need to be added with `git add`.
+
+Sometimes a change has already been applied to the tree or is a change
+requiring porting that is reverted later on. In that case, `mergify
+skip` can be used. It restores all modified files to the version in HEAD
+and then continues.
+
+#### Rolling back
+Over the course of the merge process, it is not uncommon to discover
+that a previously merged change requires additional changes. Ideally
+we'd make a fixup commit and then rebase, but `git rebase` is useless
+with merge commits. Instead, the general process is to stop the merge,
+use `git reset` to rewind to the broken commit, fix it up, and restart
+the merge. The process usually looks like
+```
+$ mergify abort
+$ git reset --hard <bad merge>
+...
+$ git commit --amend
+$ mergify start freebsd-main-<YYYYMMDD>
+```
+Sometime the commit itself isn't the problem, but rather that it
+enabled a feature than was broken with CHERI. When this happens,
+resetting to the commit before the feature was enabled, fixing things,
+and restarting can make sense.
+
+Sometimes there are too many awkward changes (e.g. one that required
+fixup in files outside the original change) and it's not worth
+backtracking to fix things in the ideal location. In that case, make
+sure to indicate that the fixup commit fixes an issue introduced in a
+particular commit to aid future users of bisect.
+
+#### Finishing up
+Once the merge is complete and basic local testing is done, it's time to
+update the `freebsd-main` branch and push the tag. (It's good to delay
+to this point in hopes of keeping `freebsd-main` relatively in sync with
+main.)
+```
+$ git push origin freebsd-main-<YYYYMMDD>:freebsd-main
```

Comment:
Sadly, this command did not work and I had to resort to checking out `freebsd-main`, merging the change, and then pushing `freebsd-main`.  This seems like maybe a bug in GH?
```suggestion
$ git checkout freebsd-main
$ git merge freebsd-main-<YYYYMMDD>
$ git push
```

Reply:
Was it complaining that `freebsd-main-YYYYMMDD` wasn't a fast-forward from `freebsd-main`?

Reply:
No, it's some other error about "bad ref" or the like.  I've lost it in my scroll back.  I'll post it when I do the next merge.

Reply:
Found the error in Slack scroll back:

```
> git push origin freebsd-main-20210305:freebsd-main
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To github.com:CTSRD-CHERI/cheribsd.git
 ! [remote rejected]           freebsd-main-20210305 -> freebsd-main (bad ref)
error: failed to push some refs to 'github.com:CTSRD-CHERI/cheribsd.git'
```

Reply:
I think this is the only thing that might need fixing before merging this in?

Reply:
Yeah, I'm still confused as to why this doesn't work. We might just want to suggest a worktree.

Reply:
I'm effectively using a worktree currently.

Reply:
I've restored the worktree based text from the original commit.